Memory improvements + Type analysis

BobVarioa opened the pull request:
Summary: 
- adds an `allocString` method to the allocator, currently only for the static allocator 
- changes `String.prototype.concat` to be a builtin method, instead of a function in codegen
- removed all `Porffor.bs/s` calls 
- adds `Porffor.allocatePage<type>` and `Porffor.allocateBytes<type>` as simple memory management utilities
- rewrites some methods to allocate bytes instead of pages
- reenables js type inference (this allows `String.prototype.concat` to function in more cases, along with many other methods)
- adds `Porffor.allocateNamedPage<type>(key)` to create a pointer to a single keyed place in memory  

Current issues: (box is checked if it is completed) 
- [x] Precompiled strings are not yet included at runtime
- [x] All strings should be initialized at the beginning of a program, and not at various locations throughout
- [x] `makeStringBuffer` should be removed, as it has been superceded by the `Porffor.allocate` methods
- [x] Improve compiler performance for precompiled strings, it's currently at unacceptable levels
- [ ] Investigate the -0.09 percent test262 drop
    - Most of these seem to be uncovered issues? i.e. there is a lot of new TypedArray failures, but that's mostly because we don't have ArrayBuffer

Future extensions:
- Avoid `Porffor.allocatePage` whenever possible, choosing `Porffor.allocateBytes` instead
- Convert the current array implementation to a linked list so that it can actually have the benefits of `allocateBytes` (set as well)
- Actual `malloc` and `free`, to be any useful requires a GC so likely far future

BobVarioa committed:
codegen: optimize string allocation

BobVarioa committed:
codegen: fix minor string bugs

BobVarioa committed:
codegen: improve string initialization code size

BobVarioa committed:
codegen: revert i64 string writing

BobVarioa committed:
codegen: allow multiple pages for strings

CanadaHonk commented on compiler/builtins/string_f64.ts:
```diff
@@ -5,4 +5,154 @@ import type {} from './porffor.d.ts';
\ No newline at end of file
+};
+
+export const __String_prototype_concat = (_this: string, arg: string) => {
```
Why is this in the same PR? Isn't it a separate thing making some new prototype methods (ty!)

BobVarioa replied:
Ah, I forgot to elaborate on that, I could just edit the previous method if you would like me to do that. But, the previous function is incorrect with my changes, mostly because it copies the entire page instead of just the length of the string.

CanadaHonk reviewed (changes requested):
Thanks! Commented

BobVarioa committed:
codegen: convert constant strings to string buffers

BobVarioa committed:
codegen: replace concat strings with function implementation

BobVarioa committed:
codegen: un-asm String.concat

BobVarioa committed:
allocators: actually allow multiple pages for strings

BobVarioa committed:
builtins/ecma262: remove unnecessary string buffer

BobVarioa committed:
builtins: seperate internal functions into a seperate file

BobVarioa committed:
builtins: replace Porffor.s/bs with actual allocations

BobVarioa committed:
builtins: rewrite some builtins to allocate memory on the fly
this also fixes some type related issues with allocations
and includes some rewrites of string to support this

BobVarioa committed:
builtins/date: cleanup Porffor internal methods

BobVarioa committed:
codegen: reenable js type analysis

BobVarioa committed:
builtins: rename porffor.ts to utils.ts
typescript doesn't like a d.ts file to have the same name as .ts apparently :(

BobVarioa committed:
codegen: move string initialization to the beginning of a program

BobVarioa committed:
builtins: eliminate some unnecessary locals

BobVarioa committed:
builtins/string: properly cast argument of includes to string

BobVarioa committed:
codegen: fix strings added at precompilation

BobVarioa committed:
builtins/string: fix string.concat
for some reason i changed the order of the args, because yes of course string concatenation is communitive

BobVarioa committed:
builtins: fix some types

BobVarioa committed:
builtins/string: rewrite string.trim

BobVarioa committed:
builtins/symbol: rewrite symbols to not use a page

BobVarioa committed:
builtins/utils: rewrite string methods to avoid looping

BobVarioa committed:
codegen: overhaul js type analysis

BobVarioa committed:
chore: merge from main

BobVarioa committed:
allocators: use map.has, not a falsy check
:facepalm:

BobVarioa committed:
builtins/string: convert compare strings into a builtin

BobVarioa committed:
builtins/array: allocate -> allocatePage

BobVarioa committed:
builtins/string: fix charcode for nan values

BobVarioa committed:
codegen/types: fix typo in prototype return type checking

BobVarioa committed:
codegen: add Porffor.allocateNamedPage

BobVarioa committed:
builtins/symbol: fix symbol implementation

BobVarioa committed:
builtins/string: revert change to string constructor

BobVarioa committed:
builtins/symbol: fix typo

BobVarioa committed:
codegen: fix nonbytestring strings

BobVarioa committed:
builtins/string: fix trim for strings

BobVarioa committed:
codegen: remove makeStringBuffer

BobVarioa commented:
I still think this PR is not quite ready to be merged, but I think it's at a stage where this is no longer a draft

BobVarioa committed:
chore: regenerate builtins

CanadaHonk reviewed (commented):
We're going to slowly incrementally add the improvements from this into main instead of just merging one chunky PR :)

BobVarioa commented on compiler/builtins/array.ts:
```diff
@@ -388,10 +388,10 @@ export const __Array_prototype_sort = (_this: any[], callbackFn: any) => {
         else {
           // 4. If comparefn is not undefined, then
           // a. Let v be ? ToNumber(? Call(comparefn, undefined, « x, y »)).
-          v = callbackFn(x, y);
```
Ah! My bad, make sense now

CanadaHonk replied:
No worries, just documenting each change, should have before anyway :)

BobVarioa commented on compiler/builtins/annexb_string.js:
```diff
@@ -4,56 +4,23 @@ export default () => {
 
   const noArgs = (a0, a1) => out += `
 export const __String_prototype_${a0} = (_this: string) => {
-  let out: string = Porffor.s\`<${a1}>\`;
```
Btw, the Porffor.s/bs calls will still have to be removed/changed once the string allocation aspects of this pr are merged
